src/tests/integration/SparkLend.t.sol

```diff
@@ -0,0 +1,400 @@
+            sparkPool.supply(address(dai), amt, address(0), 0);
+        } else {
+            // Borrow to decrease liquidity
+            uint256 amt = uint256(-deltaAmount);
+            if (amt > getLiquidity()) {
+                amt = getLiquidity();
+            }
+            sparkPool.borrow(address(dai), amt, 2, 0, address(this));
+        }
+    }
+
+    function generateInterest() internal override {
+        // Generate interest by borrowing and repaying
+        uint256 performanceBonus = daiInterestRateStrategy.performanceBonus();
+        sparkPool.supply(address(dai), performanceBonus * 4, address(this), 0);
+        sparkPool.borrow(address(dai), performanceBonus * 2, 2, 0, address(this));
+        vm.warp(block.timestamp + 1 days);
+        sparkPool.repay(address(dai), performanceBonus * 2, 2, address(this));
+        sparkPool.withdraw(address(dai), performanceBonus * 4, address(this));
+    }
+
+    function getLiquidity() internal override view returns (uint256) {
+        return dai.balanceOf(address(adai));
+    }
+
+    // --- Helper functions ---
+    function getDebtCeiling() internal view returns (uint256) {
+        (,,, uint256 line,) = dss.vat.ilks(ilk);
+        return line;
+    }
+
+    function getDebt() internal view returns (uint256) {
+        (, uint256 art) = dss.vat.urns(ilk, address(pool));
+        return art;
+    }
+
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        unchecked {
+            z = x != 0 ? ((x - 1) / y) + 1 : 0;
+        }
+    }
+
+    function getSupplyUsed(address asset) internal view returns (uint256) {
+        PoolLike.ReserveData memory data = sparkPool.getReserveData(asset);
+        return _divup((adai.scaledTotalSupply() + uint256(data.accruedToTreasury)) * data.liquidityIndex, RAY);
+    }
+
+    function getInterestRateStrategy(address asset) internal view returns (address) {
+        PoolLike.ReserveData memory data = sparkPool.getReserveData(asset);
+        return data.interestRateStrategyAddress;
+    }
+
+    function forceUpdateIndicies(address asset) internal {
+        // Do the flashloan trick to force update indicies
+        sparkPool.flashLoanSimple(address(this), asset, 1, "", 0);
+    }
+
+    function executeOperation(
+        address,
+        uint256,
+        uint256,
+        address,
+        bytes calldata
+    ) external pure returns (bool) {
+        // Flashloan callback just immediately returns
+        return true;
+    }
+
+    function getTotalAssets(address asset) internal view returns (uint256) {
+        // Assets = DAI Liquidity + Total Debt
+        PoolLike.ReserveData memory data = sparkPool.getReserveData(asset);
+        return dai.balanceOf(address(adai)) + ATokenLike(data.variableDebtTokenAddress).totalSupply() + ATokenLike(data.stableDebtTokenAddress).totalSupply();
+    }
+
+    function getTotalLiabilities(address asset) internal view returns (uint256) {
```

Comment:
Can you explain what you mean by total liabilities? Is it the amount of DAI the pool owns to all lenders and to the treasury?

Reply:
Yes exactly.

```
Total Assets = DAI in aDAI Contract (Idle Liquidity) + Stable Debt + Variable Debt
Total Liabilities = aDAI Supply (Users Lending) + Amount Owed to Treasury
```